### ISSUE TITLE
Source map support

### ISSUE THREAD

**Issue by pavelkukov** [mozilla/rhino#260]
This is not an issue, it is question.
**Did Rhino support source maps?**
If not, is it planned?

**Comment by gbrail** [comment]
There's not reason why we couldn't implement those in Rhino, but at the
same time we don't have any plans. There are only a few people working on
Rhino, and they're all working very sporadically, so any help is
appreciated and we take pull requests ;-)

With that said, as I understand it, the Closure Compiler supports source
maps, and furthermore the Closure Compiler is based on a fork of Rhino:

https://github.com/google/closure-compiler/wiki/Source-Maps

Perhaps someone from that team can help.

On Fri, Jan 29, 2016 at 5:33 AM, Pavel Kukov <notifications@github.com>
wrote:

> This is not an issue, it is question
> *Did Rhino support source maps?*
> If not, is it planned?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/rhino/issues/260>.
>



-- 
*Greg Brail* | *apigee <https://apigee.com/>* | twitter @gbrail
<http://twitter.com/gbrail> @apigee <https://twitter.com/apigee>

**Comment by pavelkukov** [comment]
Thank you @gbrail !
Very helpful info about closure compiler.

**Comment by pastelmind** [comment]
I landed here while looking for the means of enable source map-driven exception stack traces, similar to [node-source-map-support](https://github.com/evanw/node-source-map-support).

**TL;DR: source map-driven stack traces are impossible in Rhino.**

All source map-consuming tools need both the line and column numbers of the transformed code to retrieve the corresponding position from the original source. For example, Mozilla's own [source-map](https://github.com/mozilla/source-map/#sourcemapconsumerprototypeoriginalpositionforgeneratedposition) expects the `line` and `column` parameters. However, Rhino does not provide a column number, even if V8-style stack traces are enabled. [It simply prints a 0.](https://github.com/mozilla/rhino/blob/Rhino1_7_13_Release/src/org/mozilla/javascript/ScriptStackElement.java#L93)

Rhino depends heavily on Java's own stack trace mechanism to generate JavaScript stack traces. See [here](https://github.com/mozilla/rhino/blob/Rhino1_7_13_Release/src/org/mozilla/javascript/RhinoException.java#L326) where the built-in `StackTraceElement` object is used to create a `ScriptStackTrace` object. Unfortunately, [Java doesn't support column numbers either.](https://stackoverflow.com/q/23382380)

To make source maps work for stack traces, I expect we'll need to somehow attach column numbers to every JavaScript function and extract that information when constructing a stack trace.
